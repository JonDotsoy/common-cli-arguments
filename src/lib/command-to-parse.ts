import { get } from "@jondotsoy/utils-js/get";

function* parse(args: unknown[]) {
  const memoryKeys = new Set<string>();

  const uniqueKey = (value: string) => {
    let k = `${value}`;
    let findNext = 1;

    while (true) {
      if (!memoryKeys.has(k)) {
        memoryKeys.add(k);
        return k;
      }
      k = `${value}${findNext++}`;
    }
  };

  const sanitizeKeyword = (value: string) =>
    uniqueKey(
      value
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, " ")
        .trim()
        .replace(/\W(\w)/g, (_: string, char: string) => char.toUpperCase()),
    );

  for (const arg of args) {
    const literal = get.string(arg, "literal") ?? "asd";
    const isOptionsArg = get.boolean(arg, "isOptionsArg") ?? false;
    const isOptional = get.boolean(arg, "isOptional") ?? false;
    const isSpread = get.boolean(arg, "isSpread") ?? false;

    const keyword = sanitizeKeyword(literal);
    const keywordValue = isSpread ? `...${keyword}` : keyword;
    const symbol = ":";
    const value = isOptionsArg ? "T" : isSpread ? "string[]" : "string";
    const v = isOptional ? (isSpread ? value : `${value} | null`) : value;

    yield `${keywordValue}${symbol} ${v}`;
  }
}

function* parseArgsToTsBody(value: unknown) {
  for (const usageMatrix of get.array(value, "usageMatrix") ?? []) {
    const args = get.array(usageMatrix, "args") ?? [];
    yield `[${Array.from(parse(args)).join(", ")}]`;
  }
}

export const generateArgsToTs = (value: unknown) => {
  const sourcePaths = parseArgsToTsBody(value);
  return `export type Args<T> = ${Array.from(sourcePaths).join(" | ")}`;
};
