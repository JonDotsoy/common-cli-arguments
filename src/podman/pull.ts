import {
  pullHelpsMessages,
  pullVersion,
  helpMessageFilesToCommandsTsFile,
  generateCommandLib,
} from "../lib/utils";

console.time("pulling podman helps");

await pullVersion("podman");
await pullHelpsMessages("podman");
await helpMessageFilesToCommandsTsFile();
await generateCommandLib(
  await import("./commands"),
  new URL("./podman.ts", import.meta.url),
);

console.timeEnd("pulling podman helps");
