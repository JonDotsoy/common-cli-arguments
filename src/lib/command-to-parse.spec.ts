import { expect, it } from "bun:test";
import { commands } from "../podman/commands";
import { generateArgsToTs } from "./command-to-parse";

it("should generate podman diff types", () =>
  expect(generateArgsToTs(commands.podman.commands.diff)).toMatchSnapshot());
it("should generate podman types", () =>
  expect(generateArgsToTs(commands.podman)).toMatchSnapshot());
it("should generate podman attach types", () =>
  expect(generateArgsToTs(commands.podman.commands.attach)).toMatchSnapshot());
it("should generate podman volume ls types", () =>
  expect(
    generateArgsToTs(commands.podman.commands.volume.commands.ls),
  ).toMatchSnapshot());
it("should generate podman volume create types", () =>
  expect(
    generateArgsToTs(commands.podman.commands.volume.commands.create),
  ).toMatchSnapshot());
it("should generate podman run types", () =>
  expect(generateArgsToTs(commands.podman.commands.run)).toMatchSnapshot());
it("should generate podman rm types", () =>
  expect(generateArgsToTs(commands.podman.commands.rm)).toMatchSnapshot());
it("should generate podman rmi types", () =>
  expect(generateArgsToTs(commands.podman.commands.rmi)).toMatchSnapshot());
it("should generate podman build types", () =>
  expect(generateArgsToTs(commands.podman.commands.build)).toMatchSnapshot());
it("should generate podman cp types", () =>
  expect(generateArgsToTs(commands.podman.commands.cp)).toMatchSnapshot());
it("should generate podman tag types", () =>
  expect(generateArgsToTs(commands.podman.commands.tag)).toMatchSnapshot());
it("should generate podman untag types", () =>
  expect(generateArgsToTs(commands.podman.commands.untag)).toMatchSnapshot());
