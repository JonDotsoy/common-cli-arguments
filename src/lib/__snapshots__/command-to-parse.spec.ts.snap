// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate podman diff types 1`] = `"export type Args<T> = [options: T | null, containerImage: string, containerImage1: string | null]"`;

exports[`should generate podman types 1`] = `"export type Args<T> = [options: T | null, command: string | null]"`;

exports[`should generate podman attach types 1`] = `"export type Args<T> = [options: T | null, container: string]"`;

exports[`should generate podman volume ls types 1`] = `"export type Args<T> = [options: T | null]"`;

exports[`should generate podman volume create types 1`] = `"export type Args<T> = [options: T | null, name: string | null]"`;

exports[`should generate podman run types 1`] = `"export type Args<T> = [options: T | null, image: string, commandArg: string | null]"`;

exports[`should generate podman rm types 1`] = `"export type Args<T> = [options: T | null, container: string, ...container1: string[]]"`;

exports[`should generate podman rmi types 1`] = `"export type Args<T> = [options: T | null, image: string, ...image1: string[]]"`;

exports[`should generate podman build types 1`] = `"export type Args<T> = [options: T | null, context: string | null]"`;

exports[`should generate podman cp types 1`] = `"export type Args<T> = [options: T | null, containerSrcPath: string, containerDestPath: string]"`;

exports[`should generate podman tag types 1`] = `"export type Args<T> = [options: T | null, image: string, targetName: string, ...targetName1: string[]]"`;

exports[`should generate podman untag types 1`] = `"export type Args<T> = [options: T | null, image: string, ...image1: string[]]"`;
